multicluster/controllers/multicluster/member/serviceexport_controller.go

```diff
@@ -626,10 +674,155 @@ func (r *ServiceExportReconciler) updateOrCreateResourceExport(resName string,
 	return nil
 }
 
+// getSubsetsFromEndpointSlice will get all ready endpoints from all the EndpointSlices which will
+// be merged to one Endpoints. In future, we should change to track and export individual
+// EndpointSlices, rather than merge them to one Endpoints.
+func (r *ServiceExportReconciler) getSubsetsFromEndpointSlice(ctx context.Context, req ctrl.Request) ([]corev1.EndpointSubset, bool, error) {
+	epSliceList := &discovery.EndpointSliceList{}
+	hasReadyEndpoints := false
+	err := r.Client.List(ctx, epSliceList, &client.ListOptions{
+		LabelSelector: getEndpointSliceLabelSelector(req.Name),
+		Namespace:     req.Namespace})
+	if err != nil {
+		return nil, hasReadyEndpoints, err
+	}
+	if len(epSliceList.Items) == 0 {
+		return nil, hasReadyEndpoints, nil
+	}
+	var subsets []corev1.EndpointSubset
+	for _, eps := range epSliceList.Items {
+		if eps.AddressType == discovery.AddressTypeIPv4 {
+			var ports []corev1.EndpointPort
+			if r.endpointIPType == common.EndpointIPTypePodIP {
+				ports = convertEndpointPorts(eps.Ports)
+			}
+			for _, ep := range eps.Endpoints {
+				if *ep.Conditions.Ready {
+					// We only cares if there is ready Endpoints for a Service when the endpointIPType is ClusterIP,
+					// so skip handling the EndpointSubset and stop the loop early if any ready address is found.
+					if r.endpointIPType == common.EndpointIPTypeClusterIP {
+						return nil, true, nil
+					}
+					subset := corev1.EndpointSubset{}
+					readyAddresses := ipsToEndpointAddresses(ep.Addresses)
+					subset.Addresses = append(subset.Addresses, readyAddresses...)
+					subset.Ports = ports
+					subsets = append(subsets, subset)
```

Comment:
It generates one subset per Endpoint, but I think the relationship between EndpointSlice and Endpoints is one EndpointSlice maps to one subset in Endpoints?
> // EndpointSubset is a group of addresses with a common set of ports. The
// expanded set of endpoints is the Cartesian product of Addresses x Ports.

Otherwise the `ports` field is rather redundant as there is a duplicate in each subset.

Reply:
Refined.